### PR TITLE
Scripts for PSI minishift tests

### PR DESCRIPTION
**Pull request by rnapoles-rh**
**What type of PR is this?**

/kind feature

**What does does this PR do / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #?4288 

**PR acceptance criteria**:

- [ ] Unit test 

- [x] Integration test 

- [ ] Documentation 

- [x] I have read the [test guidelines](https://github.com/openshift/odo/blob/master/docs/dev/test-architecture.adoc)

**How to test changes / Special notes to the reviewer**:
test-cmd-project and test-cmd-service integration tests should pass

### CODE REVIEW

File: scripts/minishift-all-tests.sh [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+executing() {
+   set +x
+  echo -e "\n------------------------------\n${1}\n------------------------------\n"
+   set -x
+}
+
+set -ex
+
+#Export GitHub token to avoid 
+export MINISHIFT_GITHUB_API_TOKEN="03d362b0f2a7ecf02b30bfbdd40fe720a4d6b8b3"
```
Comment (kadel) [inline scripts/minishift-all-tests.sh:12]: This should not be there. It is a big security hole! 
It needs to use some kind of mechanism for securing secrets.

Do we even need this? If I remember correctly this token is only used by minishift to check for new versions and it is not required for any other functionality.

File: scripts/minishift-all-tests.sh [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+executing "Setting up environment..."
+
+export PATH="$PATH:/usr/local/go/bin/"
+export GOPATH=$HOME/go
+git clone https://github.com/openshift/odo.git openshift/odo
```
Comment (kadel) [inline scripts/minishift-all-tests.sh:17]: This is cloning the master branch. Shouldn't it clone PR branch?
Also isn't this take care of by CI itself?

File: scripts/minishiftci-presubmit-all-test.sh [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+export CI_FIREWALL_VERSION="valpha"
+
+echo "Getting ci-firewall, see https://github.com,/mohammedzee1000/ci-firewall"
+# show commands
+set -x
+
+curl -kLO https://github.com/mohammedzee1000/ci-firewall/releases/download/valpha/ci-firewall-linux-amd64.tar.gz
+tar -xzf ci-firewall-linux-amd64.tar.gz
+
+./ci-firewall request --runscript scripts/kubernestes-all-test.sh --timeout 2h15m
```
Comment (kadel) [inline scripts/minishiftci-presubmit-all-test.sh:31]: shouldn't it call `minishift-all-tests.sh` and not `kubernestes-all-test.sh`?

File: scripts/minishift-all-tests.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+cd openshift/odo
+
+mkdir -p $GOPATH/bin
+make goget-ginkgo
+export PATH="$PATH:$(pwd):$GOPATH/bin"
+
+executing "Building ODO..."
+make bin
+sudo cp odo /usr/bin
```
Comment (anandrkskd) [inline scripts/minishift-all-tests.sh:25]: why are we copying it in "/usr/bin", shouldn't we use "$GOPATH/bin"?

File: scripts/minishiftci-presubmit-all-test.sh (hunk 3 above)
Comment (rnapoles-rh) [reply scripts/minishiftci-presubmit-all-test.sh:31]: you are right, updating it

File: scripts/minishift-all-tests.sh (hunk 2 above)
Comment (mohammedzee1000) [reply scripts/minishift-all-tests.sh:16]: Yes you don't have to clone. ci-firewall will already clone repo and check out the pr

File: scripts/minishift-all-tests.sh (hunk 4 above)
Comment (rnapoles-rh) [reply scripts/minishift-all-tests.sh:25]: $GOPATH/bin holds go binaries. In general all applications go to /usr/bin. 
The same is done for minikube

File: scripts/minishift-all-tests.sh (hunk 2 above)
Comment (rnapoles-rh) [reply scripts/minishift-all-tests.sh:16]: removing the cloning

File: scripts/minishift-all-tests.sh (hunk 4 above)
Comment (anandrkskd) [reply scripts/minishift-all-tests.sh:25]: If we are creating a fresh vm every time we test then it makes sense but if we using a long running cluster we should use  "$GOPATH/bin".
what do you think?

Comment (rnapoles-rh) [reply scripts/minishift-all-tests.sh:25]: agreed, I will make the change

**Comment by rnapoles-rh** [comment]
the devfile debug test for java-openliberty faied for this PR, however for PR4437 nodejs failed and java-openliberty passed. rerrunning.

**Comment by rnapoles-rh** [comment]
Test execution goes over the 2 hour timeout

File: scripts/delete-cluster-projects.sh [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+set -ex
+
+shout "| Cleaning up projects from cluster ..."
+
+shout "| Logging in to minishift..."
+
+oc login -u developer -p developer --insecure-skip-tls-verify $(minishift ip):8443
+PROJECT_LIST=$(oc get projects)
+
+for i in $PROJECT_LIST; do
```
Comment (kadel) [inline scripts/delete-cluster-projects.sh:18]: Why is this needed?

File: scripts/minishift-execute-tests.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+set -ex
+
+eval $(minishift oc-env)
+
+shout "| Logging in to minishift..."
+oc login -u developer -p developer --insecure-skip-tls-verify $(minishift ip):8443
+
+shout "| Executing: generic, login, component command and plugin handler integration tests"
+make test-generic
```
Comment (kadel) [inline scripts/minishift-execute-tests.sh:17]: Can we consolidate test execution into one file together with https://github.com/openshift/odo/pull/4401/files and tests that are executed against our PSI OCP cluster and also Prow?

The list of tests is stored in multiple places, and it will be almost impossible to remember all the places where new test need to be added :-( Having it in one place would help a lot.

File: scripts/minishiftci-presubmit-all-test.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
```
Comment (kadel) [inline scripts/minishiftci-presubmit-all-test.sh:1]: I made the same comment at  https://github.com/openshift/odo/pull/4401#discussion_r588263997

> This script looks almost the same as what we already have in https://github.com/openshift/odo/blob/master/scripts/openshiftci-e2e-4x-psi-tests.sh
> 
> We should reuse what we already have and not create duplication. If needed the existing script should be generalized or extracted into a function.

This makes it 3rd instance of almost the same code.

**Review by kadel — CHANGES_REQUESTED** [review]

File: scripts/minishift-execute-tests.sh (hunk 6 above)
Comment (kadel) [reply scripts/minishift-execute-tests.sh:17]: Or we could simply create `make test-all` and use that everywhere

File: scripts/delete-cluster-projects.sh (hunk 5 above)
Comment (rnapoles-rh) [reply scripts/delete-cluster-projects.sh:18]: This was to cleanup any projects left behind in the cluster, however I will remove it for now as we can approach the "no proper cleanup by tests" problem in a better way.

File: scripts/minishift-execute-tests.sh (hunk 6 above)
Comment (rnapoles-rh) [reply scripts/minishift-execute-tests.sh:17]: Sure, I will consolidate in one script

File: scripts/minishiftci-presubmit-all-test.sh (hunk 7 above)
Comment (rnapoles-rh) [reply scripts/minishiftci-presubmit-all-test.sh:1]: I will consolidate into one as well.

File: scripts/minici-presubmit-all-tests.sh [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+tar -xzf ci-firewall-linux-amd64.tar.gz
+
+case $1 in
+    minikube)
+        ./ci-firewall request --sendqueue amqp.ci.queue.minikube.send --sendtopic amqp.ci.topic.minikube.send --runscript scripts/kubernetes-all-test.sh  --timeout 2h15m
+        ;;
+    minishift)
+        ./ci-firewall request --sendqueue amqp.ci.queue.minishift.send --sendtopic amqp.ci.topic.minishift.send --setupscript minishift-setup-env.sh --runscript scripts/minishift-execute-test.sh  --timeout 4h00m
+        ;;
+esac
```
Comment (anandrkskd) [inline scripts/minici-presubmit-all-tests.sh:49]: Can't we use variable to define sendqueue and other flags as we are using switch case, by this we will make the command more generic and then can remove this switchcase. 
does that make sense?

File: scripts/minishift-execute-tests.sh [hunk 9]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env bash
```
Comment (anandrkskd) [inline scripts/minishift-execute-tests.sh:1]: Just being curious, we are going to merge this script with minikube script to test. right?
And we will use this script to setup and run the test for minikube and minishift? 
/cc: @rnapoles-rh

**Review by anandrkskd — COMMENTED** [review]
There are some question, please check comments

File: scripts/minici-presubmit-all-tests.sh (hunk 8 above)
Comment (rnapoles-rh) [reply scripts/minici-presubmit-all-tests.sh:49]: @anandrkskd there is a dependency on the kubernetesci-presubmit-all-tests.sh script being handled by Priti, to avoid clashing I decided to keep it this way for now, once we agree to replace decommission kubernetesci-presubmit-all-tests.sh I can use variables as you suggest.

File: scripts/minishift-all-tests.sh (hunk 4 above)
Comment (rnapoles-rh) [reply scripts/minishift-all-tests.sh]: change made

File: scripts/minici-presubmit-all-tests.sh (hunk 8 above)
Comment (rnapoles-rh) [reply scripts/minici-presubmit-all-tests.sh]: I created a presubmit script that handles both minikube and minishift to be able to define variables for sendqueue and sendtopic

File: scripts/minishift-execute-tests.sh (hunk 9 above)
Comment (rnapoles-rh) [reply scripts/minishift-execute-tests.sh]: yes, done

**Comment by rnapoles-rh** [comment]
the ci/prow/v4.7-e2e-4x-ps failures are not related to the code being delivered

File: scripts/minici-presubmit-all-tests.sh [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+##### ci-firewall parameters end
+
+# The version of CI_FIREWALL TO USE
+export CI_FIREWALL_VERSION="valpha"
+
+echo "Getting ci-firewall, see https://github.com,/mohammedzee1000/ci-firewall"
+# show commands
+set -x
+
+curl -kLO https://github.com/mohammedzee1000/ci-firewall/releases/download/valpha/ci-firewall-linux-amd64.tar.gz
```
Comment (anandrkskd) [inline scripts/minici-presubmit-all-tests.sh:49]: shouldn't we use CI_FIREWALL_VERSION variable in url?

Comment (rnapoles-rh) [reply scripts/minici-presubmit-all-tests.sh]: You are right, updated

**Review by anandrkskd — APPROVED** [review]
/lgtm
